# Imports

In [1]:
import pandas as pd

Data Import

In [2]:
df = pd.read_csv("dirty_financial_transactions.csv")

In [3]:
df.head()

,Transaction_ID,Transaction_Date,Customer_ID,Product_Name,Quantity,Price,Payment_Method,Transaction_Status
0,T0001,2024-08-02,C2205,Headphones,-5.0,$420.21,pay pal,NaN
1,T0002,2020-02-10,C3156,Coffee,469.0,-445.34202525395585,creditcard,Pending
2,T0003,2025-02-30,C2919,Tablet,-4.0,810.9930123946459,credit card,completed
3,T0004,2020-08-17,C3009,Tab,-7.0,868.6083413217348,PayPal,Pending
4,T0005,2025-02-30,C3488,Coffee Machine,-10.0,-763.1224490039416,PayPal,completed


In [4]:
df.sample(10)

,Transaction_ID,Transaction_Date,Customer_ID,Product_Name,Quantity,Price,Payment_Method,Transaction_Status
30815,T30816,2023-13-01,C3804,Smartphone,-2.0,NaN,pay pal,Pending
81403,T81404,2023-13-01,C1199,Laptop,5.0,658.6058071281445,Credit Card,Failed
55207,T55208,2023-13-01,C4044,Headphones,596.0,NaN,credit card,NaN
79684,T79685,NaN,C757,Smartphone,882.0,NaN,Credit Card,NaN
60400,T60401,2024-06-09,C560,Coffee Machine,332.0,272.1281709042265,creditcard,complete
78575,T78576,2023-13-01,C120,Laptop,5.0,749.0869316055105,Credit Card,Completed
85084,T85085,2023-10-24,C870,Coffee Machine,476.0,-106.29272165184864,creditcard,NaN
53821,NaN,2023-13-01,C4181,Laptop,-8.0,719.5205343359293,pay pal,Completed
42869,T42870,2023-13-01,C2278,Headphones,-7.0,NaN,PayPal,Completed
68394,T68395,2023-13-01,C1369,Tablet,578.0,860.8486214177483,pay pal,Completed


In [5]:
df.info(memory_usage='deep')

<class 'pandas.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 8 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   Transaction_ID      94982 non-null   str    
 1   Transaction_Date    95120 non-null   str    
 2   Customer_ID         95122 non-null   str    
 3   Product_Name        100000 non-null  str    
 4   Quantity            94981 non-null   float64
 5   Price               66503 non-null   str    
 6   Payment_Method      100000 non-null  str    
 7   Transaction_Status  83321 non-null   str    
dtypes: float64(1), str(7)
memory usage: 37.6 MB


In [6]:
df.describe()

,Quantity
count,94981.000000
mean,183.883914
std,299.292365
min,-10.000000
25%,-3.000000
50%,6.000000
75%,327.000000
max,1000.000000


In [7]:
print(f"Missing Value Counts\n{df.isnull().sum()}\n")
print(f"Missing Value Counts : Percentage\n{df.isnull().sum() / len(df) * 100}\n")

Missing Value Counts
Transaction_ID         5018
Transaction_Date       4880
Customer_ID            4878
Product_Name              0
Quantity               5019
Price                 33497
Payment_Method            0
Transaction_Status    16679
dtype: int64

Missing Value Counts : Percentage
Transaction_ID         5.018
Transaction_Date       4.880
Customer_ID            4.878
Product_Name           0.000
Quantity               5.019
Price                 33.497
Payment_Method         0.000
Transaction_Status    16.679
dtype: float64



In [8]:
print(f"Total Missing Value Counts\n{df.isnull().sum().sum()} Out of {len(df)*len(df.columns)}\n")
print(f"Total Missing Value Counts : Percentage : {df.isnull().sum().sum() / (len(df)*len(df.columns)) * 100}\n")

Total Missing Value Counts
69971 Out of 800000

Total Missing Value Counts : Percentage : 8.746375



# Findings
* There are 100k rows and 8 columns in this dataset.
* Mismatch of dtypes for multiple columns.
* The should be numerical columns cannot be easily converted due the presence of string values in it, i.e 'Price' column where in some places the $ sign in present, some price values are even in negative, this also needs to be taken care of.
* Solutin for 'Price' column : After we remove the $ character and fix the negative sign issues, to fill the NaN gap we can find out the average unit price of each product categroy and then put those, then we will be able to bring back most of the data, and i feel like it's better than putting any all-over mean or average values in there, since the quantity is also a factor not to be ignored.
* The 'Quantity' column too has the negative sign problem along with some null data.
* Solution for 'Quantity' column : We can use the same method as of 'Price' column, by finding the average quantity value for each product type and imputing by that.
* "Product_Name" column has the problem of incorrect or incomplete names of products, this can be solved by using the Regular Expression lib.
* Column "Payment_Method" and "Transaction_Status" have the same naming problem as "Product_Name", and it can be solved using the regular expression lib, better convert the texts to lowercase for uniformity. It would be better to put new features based on payment methods, like putting short forms i.e "CC" instead of "credit card", "C" instead of "cash", etc. same for the "Payment_Status"" column, its better to label encode the status.
* "Transaction_Date" has a problem with incorrect date format and missing data, we can convert it into datetime format and fill the missing values by forward filling or backward filling.
* "Transaction_ID" and "Customer_ID" columns can be dropped if this data is going to be fed into any ML model, as they are not required for prediction, but if this data is for any sales team then these columns should be kept for unsupervised prediction purpose.
* After the cleaning of the dataset, the clean data must be label encoded (since there are more columns with categorical values) and scaled(price and quantity number digits vary a lot).

## Seconday Solution Investigation : Dropping all null containing rows

In [9]:
df_no_null = df.dropna(axis=0)
df_no_null.sample(5)

,Transaction_ID,Transaction_Date,Customer_ID,Product_Name,Quantity,Price,Payment_Method,Transaction_Status
21560,T21561,2023-09-06,C4338,Headphones,301.0,$405.90,Credit Card,Failed
67330,T67331,2025-02-30,C1331,Headphones,879.0,$-274.61,Cash,complete
43140,T43141,2023-13-01,C2139,Tablet,773.0,212.3173863134379,creditcard,Completed
31291,T31292,2020-03-12,C150,Coffee Machine,500.0,-392.57692449748845,PayPal,Failed
42199,T42200,2025-02-30,C3591,Smartphone,-1.0,-878.1916491152006,PayPal,Pending


In [10]:
df_no_null.info(memory_usage='deep')

<class 'pandas.DataFrame'>
Index: 45333 entries, 1 to 99999
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Transaction_ID      45333 non-null  str    
 1   Transaction_Date    45333 non-null  str    
 2   Customer_ID         45333 non-null  str    
 3   Product_Name        45333 non-null  str    
 4   Quantity            45333 non-null  float64
 5   Price               45333 non-null  str    
 6   Payment_Method      45333 non-null  str    
 7   Transaction_Status  45333 non-null  str    
dtypes: float64(1), str(7)
memory usage: 18.2 MB


# Findings
* Removing all null values results in huge loss of information, about 55% of info loss, which means based on previous percentages of missing data found which totals 67% of null values, we can say that only 12% of null values exist in common rows, others are independent.
* Therefore this approach is not at all fruitfull and should be avoided.
* If dropping by null values, then considering the columns with least null values or solid logic in this domain will help to simplify the process a bit.